File: C/20.观察者模式/test/ObserverPatternDemo.c
/**
 * @file ObserverPatternDemo.c
 * @author {Layne} ({shu_huanglei@163.com})
 * @brief
 * @version 0.1
 * @date 2020-08-09
 *
 * @copyright Copyright (c) 2020
 *
 */
#include <stdio.h>

#include "Subject.h"

int main(int argc, char const* argv[]) {
    Subject* subject = subject_create();

    Observer* hexa_observer = hexa_observer_create(subject);
    Observer* octal_observer = octal_observer_create(subject);
    Observer* binary_observer = binary_obbserver_create(subject);

    printf("First state change: 15\n");
    subject_set_state(subject, 15);
    printf("Second state change: 10\n");
    subject_set_state(subject, 10);

    subject_destroy(&subject);
    observer_destroy(&hexa_observer);
    observer_destroy(&octal_observer);
    observer_destroy(&binary_observer);
    return 0;
}

File: Objective-C/20.ObserverPattern/ObserverPatternDemo.m
//
//  ObserverPatternDemo.m
//  20.ObserverPattern
//
//  Created by Layne on 2020/11/13.
//

#import "BinaryObserver.h"
#import "HexaObserver.h"
#import "Observer.h"
#import "OctalObserver.h"
#import "Subject.h"

int main(int argc, const char *argv[]) {
    @autoreleasepool {
        Subject *subject = [[Subject alloc] init];
        HexaObserver *hexaObserver = [[HexaObserver alloc] initWithSubject:subject];
        OctalObserver *octalObserver = [[OctalObserver alloc] initWithSubject:subject];
        BinaryObserver *binaryObserver = [[BinaryObserver alloc] initWithSubject:subject];

        NSLog(@"First state change: 15");
        [subject setState:15];

        NSLog(@"Second state change: 10");
        [subject setState:10];
    }
    return 0;
}
